Mathlib/Tactic/ApplyRules.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2022 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Util.TermUnsafe
+import Mathlib.Tactic.Repeat
+
+/-!
+# The `apply_rules` tactic
+
+The `apply_rules` tactic calls `apply` (with a specified set of lemmas) and `assumption`
+repeatedly, until no more applications are possible.
+-/
+
+
+namespace Mathlib.Tactic
+open Lean Meta Elab Tactic Term Parser.Tactic
+
+/--
+`applyFirst l`, for `l : List (TacticM Expr)`,
+tries to apply one of the lemmas generated by the tactics in `l` to the first goal, and
+fail if none succeeds.
+
+Implementation note:
+We pass a list of `TacticM Expr` rather than just `Expr`,
+because when `(← getMainGoal).apply e` succeeds, it can modify `e` via unification,
+blocking subsequent applications which would succeed but unify differently.
+This implementation ensures that the expressions are freshly generated for each
+invocation of `apply`.
+-/
+def applyFirst (cfg : ApplyConfig) (L : List (TacticM Expr)) : TacticM Unit := withMainContext do
+  let g ← getMainGoal
+  setGoals $ ← L.firstM (fun t => do g.apply (← t) cfg)
+
+/--
+Implementation of the `apply_rules` tactic.
+-/
+def applyRules (bound : Nat) (cfg : ApplyConfig) (lemmas : List (TacticM Expr)) : TacticM Unit :=
+  repeatAllSubgoals bound (do (evalTactic $ ← `(tactic|assumption)) <|> do _ ← applyFirst cfg lemmas)
```

Comment:
I don't think we want to encourage combinators over `TacticM`, in favor of implementations on `MetaM` directly where possible.

Reply:
To clarify: you're saying here I should not write `repeatAllSubgoals` in `TacticM`?

Reply:
You can also write it parametric over the monad; there are a lot of combinators written like that. I'm not yet sure what you need for applyRules specifically